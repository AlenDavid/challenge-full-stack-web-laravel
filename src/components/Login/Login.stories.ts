import Login from "./Login.vue"

export default {
	title: "Components/Login",
	component: Login,
}

const Template = (args: any, { argTypes }: any) => ({
	props: Object.keys(argTypes),
	components: { Login },
	template: "<login v-bind='$props' />",
})

export const Default: any = Template.bind({})

export const Filled: any = Template.bind({})
Filled.args = {
	email: "example@email.com",
	password: "password",
}

export const EmailInvalid: any = Template.bind({})
EmailInvalid.args = {
	email: "invalid as email",
}
