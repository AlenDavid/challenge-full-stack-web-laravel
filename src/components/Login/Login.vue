<template>
	<v-container>
		<v-form ref="ref" lazy-validation v-model="valid">
			<v-text-field
				label="E-mail"
				type="email"
				v-model="email"
				:rules="emailRules"
				required
			></v-text-field>
			<v-text-field
				label="Password"
				type="password"
				v-model="password"
				:rules="passwordRules"
				required
			></v-text-field>
			<v-btn
				x-large
				block
				:disabled="!valid"
				@click="onLogin"
				color="primary"
				:loading="loading"
				>Login</v-btn
			>
		</v-form>
		<v-snackbar top v-model="snackbarShow" :timeout="3000">
			{{ snackbarMessage }}
			<template v-slot:action="{ attrs }">
				<v-btn color="blue" text v-bind="attrs" @click="snackbarShow = false">
					Close
				</v-btn>
			</template>
		</v-snackbar>
	</v-container>
</template>

<script lang="ts">
import router from "@/router"
import user from "@/store/modules/user"
import Vue from "vue"

export default Vue.extend({
	name: "Login",
	data: () => ({
		ref: {},
		email: "",
		password: "",
		valid: true,
		emailRules: [
			(email: string) => !!email || "E-mail is required.",
			(email: string) =>
				/^([\w\d]+)@([\w\d]+.)+[\w\d.]+([\w\d])+$/.test(email) ||
				"E-mail is invalid.",
		],
		passwordRules: [
			(password: string) => !!password || "Password is required.",
			(password: string) =>
				password.length > 0 || "Password must not be empty.",
		],
		loading: false,
		snackbarShow: false,
		snackbarMessage: "",
	}),
	methods: {
		validate() {
			// eslint-disable-next-line
			this.valid = (this.$refs.ref as any).validate()
			return this.valid
		},
		onLogin() {
			// only do login if form is valid
			if (this.validate()) {
				this.loading = true
				const body = new FormData()

				body.set("email", this.email)
				body.set("password", this.password)

				fetch("/api/login", {
					method: "POST",
					body,
				})
					.then((res) => res.json())
					// eslint-disable-next-line
					.then(({ data, error }: any) => {
						if (error) {
							this.snackbarMessage = error[0] || error
							this.snackbarShow = true
						}

						if (data) {
							// save login information
							return new Promise(() =>
								setTimeout(() => {
									router.push("/")
									user.commit("login", data)
								}, 1500)
							)
						}
					})
					.finally(() => {
						this.loading = false
					})
			}
		},
	},
})
</script>
