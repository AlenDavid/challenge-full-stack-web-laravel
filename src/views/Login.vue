<template>
  <v-container>
    <div>Login page</div>
  </v-container>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "Login",
});
</script>
