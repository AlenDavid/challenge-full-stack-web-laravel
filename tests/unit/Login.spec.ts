import { shallowMount } from "@vue/test-utils";

import Login from "@/components/Login/Login.vue";

describe("Login.vue", () => {
  it("should match snapshot component", () => {
    const wrapper = shallowMount(Login);

    expect(wrapper).toMatchSnapshot();
  });

  it("should see email and password", () => {
    const wrapper = shallowMount(Login, {
      propsData: {
        email: "example@example.com",
        password: "password",
      },
    });

    // select the text field
    const emailInput = wrapper.find("[type='email']");
    const passInput = wrapper.find("[type='password']");

    expect(emailInput.attributes().value).toMatch("example@example.com");
    expect(passInput.attributes().value).toMatch("password");
  });

  it("should see login button", () => {
    const wrapper = shallowMount(Login, {
      propsData: {
        email: "example@example.com",
        password: "password",
      },
    });

    // select the login button
    const button = wrapper.find("[type='submit']");

    expect(button.html()).toMatch("Login");
  });
});
