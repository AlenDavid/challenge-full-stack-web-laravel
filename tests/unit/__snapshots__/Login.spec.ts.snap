// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue should match snapshot component 1`] = `
<v-container-stub tag="div">
  <v-form-stub lazyvalidation="true" value="true">
    <v-text-field-stub errorcount="1" errormessages="" messages="" rules="email => !!email || &quot;E-mail is required.&quot;,email => _constants.EMAIL_PATTERN.test(email) || &quot;E-mail is invalid.&quot;" successmessages="" value="" backgroundcolor="" label="E-mail" loaderheight="2" clearicon="$clear" type="email" required=""></v-text-field-stub>
    <v-text-field-stub errorcount="1" errormessages="" messages="" rules="password => !!password || &quot;Password is required.&quot;,password => password.length > 0 || &quot;Password must not be empty.&quot;" successmessages="" value="" backgroundcolor="" label="Password" loaderheight="2" clearicon="$clear" type="password" required=""></v-text-field-stub>
    <v-btn-stub color="primary" tag="button" activeclass="" xlarge="true" block="true" type="button" test-id="login-button">Login</v-btn-stub>
  </v-form-stub>
  <v-snackbar-stub tag="div" top="true" contentclass="" timeout="3000" transition="v-snack-transition">

  </v-snackbar-stub>
</v-container-stub>
`;
